Fix ieee802154 associations

miquelraynal opened the pull request:
Hello all,

I'm currently working on improving the Linux kernel ieee802154 stack and right now I'm working on getting various MLME operations working. I started with simulators (mac802154_hwsim), I then tested the changes with atusb hardware and I decided to do the extra mile by validating my work against Zephyr stack with an Arduino Nano BLE. My workflow involved doing:
```
ieee802154 set_chan 13
net iface up 1
ieee802154 ack 1
ieee802154 scan active 13 1000
ieee802154 associate 2 1
```

The least I can say, is that both sides needed fixes, so here are all the changes I gathered in my branch. In these fixes the first six are meant to be upstreamed, but the last two are quick and dirty fixes against Zephyr internal notifiers which I completely failed to get up and working. I really did not understand these two issues, maybe I missed a configuration option or something like that, so any advice about how to make it work would be appreciated because on my side, it does not.

Cheers,
Miquèl

miquelraynal committed:
net: l2: ieee802154: Change the hardware filters while associating

During an association, the peer we are trying to connect to will send us
an association response frame with the destination PAN ID set to the PAN
ID we try to join. If we do not update the hardware address filters
beforehands, it is likely that the hardware will just discard the
response and the association will fail.

Signed-off-by: Miquel Raynal <miquel.raynal@bootlin.com>

miquelraynal committed:
net: l2: ieee802154: Fix a wrong channel change during association

The current association logic starts by changing the channel. The way it
is done is wrong because it dereferences req->channel which is simply
not initialized by the caller. But anyway, the command itself does not
support providing a channel so we must already be on the right one when
trying to associate. Hence, drop this channel change call.

Signed-off-by: Miquel Raynal <miquel.raynal@bootlin.com>

miquelraynal committed:
net: l2: ieee802154: Fix address length macros

When validating a MAC command, there is sometimes a misunderstanding of
what "src" and "dst" length mean.

There are actually two fields in the MHR:
- One giving the type of address, if it is short or extended, it is the
  value provided by the macros IEEE802154_ADDR_MODE_{SHORT,EXTENDED} and
  their respective decimal values are 2 and 3.
- One giving the size of the address field, this is
  IEEE802154_{SHORT,EXT}_ADDR_LENGTH and their value is actually 2 and 8
  (bytes).

The function validate_mac_command() provides inputs to
validate_mac_command_cfi_to_mhr() which expects the former information.

Signed-off-by: Miquel Raynal <miquel.raynal@bootlin.com>

miquelraynal committed:
net: l2: ieee802154: Avoid dangling MAC command bits

While the packet structures seem to always be reset when they are
allocated, it's apparently not the case of the data buffers. Indeed,
these are allocated differently and just attached to the packet
structure through a frag/buffer member.

Experience shows that we may get uninitialized buffers so let's set
all MAC parameter bits one by one, even the reserved ones.

Signed-off-by: Miquel Raynal <miquel.raynal@bootlin.com>

miquelraynal committed:
net: l2: ieee802154: AR shall be provided in association requests

The spec clearly states: "association requests shall set the AR bit".

Even though Zephyr can currently only implement RFD devices which are
not expected to support incoming association requests, because this MAC
command is actually processed until being voluntarily ignored, let's
ensure the expected "ar" value is right to avoid failing because of a
wrong reason.

Signed-off-by: Miquel Raynal <miquel.raynal@bootlin.com>

miquelraynal committed:
net: l2: ieee802154: Fix logic with address types

When validating a MAC command, the "src" and "dst" fields may be set to:
- IEEE802154_ADDR_MODE_SHORT (0x2)
- IEEE802154_ADDR_MODE_EXTENDED (0x3)
- IEEE802154_ADDR_MODE_SHORT | IEEE802154_ADDR_MODE_EXTENDED (0x3)
Hence when the mode check happens, any times the mode is set to SHORT
the check will fail while in practice it was meant to be valid because
the check is:

    if (src_mode == src || dst_mod == dst)

Use bitfields when relevant so that when checking capabilities we use
the bit offsets rather than the plain numbers.

Signed-off-by: Miquel Raynal <miquel.raynal@bootlin.com>

jukkar commented:
> I really did not understand these two issues, maybe I missed a configuration option or something like that, so any advice about how to make it work would be appreciated because on my side, it does not.

At least for the scan notifier issue, you need to enable
```
CONFIG_NET_L2_IEEE802154_MGMT
CONFIG_NET_MGMT
CONFIG_NET_MGMT_EVENT
```
The first one selects the NET_MGMT and NET_MGMT_EVENT automatically but make sure those three are enabled.

jukkar commented:
About the network queue bypass, at least the packet is not queued if the interface is marked down, although this might not be the case here because you can send the packet by calling `ieee802154_radio_send()` directly. There could be also other issues with the system, but you would need to enable some more debugging in order to see why the packet is not being sent out.
Could the issue be related to your other changes, does the packet TX work properly without your patches?

miquelraynal commented:
> At least for the scan notifier issue, you need to enable
> 
> ```
> CONFIG_NET_L2_IEEE802154_MGMT
> CONFIG_NET_MGMT
> CONFIG_NET_MGMT_EVENT
> ```

Indeed I forgot to provide information about the defconfig, the above three config options were enabled.

About your second question on the net queue, I'll check in the coming days.

tbursztyka commented on subsys/net/l2/ieee802154/ieee802154_frame.c:
```diff
@@ -330,7 +330,7 @@ validate_mac_command(struct ieee802154_mpdu *mpdu, uint8_t *buf, uint8_t *length
 	case IEEE802154_CFI_ORPHAN_NOTIFICATION:
 		comp = 1U;
-		src = IEEE802154_EXT_ADDR_LENGTH;
+		src = IEEE802154_ADDR_MODE_EXTENDED;
```
What about line  345 ?

miquelraynal replied:
Oh right, should be changed as well. I'll do it.

While looking again at this file, the logic really seems broken because sometimes we do:
`dst = IEEE802154_ADDR_MODE_SHORT | IEEE802154_ADDR_MODE_EXTENDED;`
Which is the equivalent currently of:
`dst = 0x2 | 0x3;`
It won't work with short addressing. I'll fix it in another patch.

tbursztyka commented on subsys/net/l2/ieee802154/ieee802154_frame.c:
```diff
@@ -291,6 +291,7 @@ validate_mac_command(struct ieee802154_mpdu *mpdu, uint8_t *buf, uint8_t *length
 	case IEEE802154_CFI_UNKNOWN:
 		return false;
 	case IEEE802154_CFI_ASSOCIATION_REQUEST:
+		ar = 1U;
```
commit message has a type:  volontarily instead of volutarily. (or willingly, either way).

tbursztyka commented on subsys/net/l2/ieee802154/ieee802154_mgmt.c:
```diff
@@ -260,6 +260,7 @@ static int ieee802154_associate(uint32_t mgmt_request, struct net_if *iface,
 	}
 
 	cmd = ieee802154_get_mac_command(pkt);
+	memset(&cmd->assoc_req.ci, 0, sizeof(cmd->assoc_req.ci));
```
just set the 2 last missing fields to 0 (the reserved ones), it will be leaner since all the other fields are set anyway.

tbursztyka reviewed (changes requested):
- Remove the 2 last commits, if there is an issue with the notifier, let's figure it out
- In all patches, remove all the "Fixes:" annotations. They do not target any github issue. (unless this has changed lately - git log tend to say the contrary - we do not follow this rule of indexing original commit

miquelraynal committed:
net: l2: ieee802154: Add missing scan callback registration

Initializing an event callback with net_mgmt_init_event_callback() just
sets some of the callback fields but do not propagate those masks in the
core. If we want to use the callback, it is necessary to also call
net_mgmt_add_event_callback().

Signed-off-by: Miquel Raynal <miquel.raynal@bootlin.com>

miquelraynal commented:
Looks like the scan event helper was initialized but not "added" to the list of events, which prevented scan results to be captured. About the bypass of the net stack with a direct radio_send(), I am not able to reproduce the error, so everything looks fine, I'll propose a new version of the series soon.

rlubos commented on subsys/net/l2/ieee802154/ieee802154_mgmt.c:
```diff
@@ -430,6 +428,9 @@ static int ieee802154_associate(uint32_t mgmt_request, struct net_if *iface,
 	}
 
 out:
+	if (ret < 0)
+		ieee802154_filter_pan_id(iface, 0);
```
Zephyr requires surrounding `{ }` even for single-line blocks (weird that checkpatch did not catch this).
```suggestion
	if (ret < 0) {
		ieee802154_filter_pan_id(iface, 0);
	}
```

miquelraynal replied:
Oh, right, I'm sorry I'm used to the Linux kernel coding style where this is not necessary, I've been warned by checkpatch somewhere else about this but I missed this one.

rlubos reviewed (commented):
Sorry for the delay, somehow lost the PR from my radar. Looks good, just one nit from my side.

rlubos commented:
@miquelraynal Please squash the fix into 24292405865e75d568a001526b615852031a3467

miquelraynal commented:
@rlubos Rebased on latest main, and squashed the commits (I messed up with Github interface, so I did squash the files manually).

rlubos reviewed (dismissed):
Thanks! LGTM

miquelraynal commented:
Hello @tbursztyka, I don't know why Github is still waiting for me to address your comments, I cannot "click" to say it's done but for me the PR is ready.
